src/generic/AhoCorasick.jl

```diff
@@ -0,0 +1,201 @@
+function ==(m1::AhoCorasickMatch, m2::AhoCorasickMatch)
+    return m1.last_position == m2.last_position &&
+           m1.keyword_index == m2.keyword_index &&
+           m1.keyword == m2.keyword
+end
+
+function AhoCorasickAutomaton(keywords::Vector{Word})
+    automaton = AhoCorasickAutomaton([], [], [])
+    construct_goto!(automaton, keywords)
+    construct_fail!(automaton)
+    return automaton
+end
+
+function aho_corasick_automaton(keywords::Vector{Word})
+    return AhoCorasickAutomaton(keywords)
+end
+
+function lookup(automaton::AhoCorasickAutomaton, current_state::Int, next_letter::Int)
+    ret_value = get(automaton.goto[current_state], next_letter, nothing)
+    if current_state == 1 && isnothing(ret_value)
+        return 1
+    end
+    return ret_value
+end
+
+
+function Base.length(automaton::AhoCorasickAutomaton)
+    return length(automaton.goto)
+end
+
+function new_state!(automaton)
+    push!(automaton.goto, Dict{Int,Int}())
+    push!(automaton.output, (typemax(Int), []))
+    push!(automaton.fail, 1)
+    return length(automaton.goto)
+end
+
+function enter!(automaton::AhoCorasickAutomaton, keyword::Word, current_index)
+    current_state = 1
+    for c in keyword
+        new_state = get!(automaton.goto[current_state], c) do
+            new_state!(automaton)
+        end
+        current_state = new_state
+    end
+    if automaton.output[current_state][1] > current_index
+        automaton.output[current_state] = (current_index, keyword)
+    end
+end
+
+function construct_goto!(automaton::AhoCorasickAutomaton, keywords::Vector{Word})
+    new_state!(automaton)
+    for (current_index, keyword) in enumerate(keywords)
+        enter!(automaton, keyword, current_index)
+    end
+end
+
+function construct_fail!(automaton::AhoCorasickAutomaton)
+    q = Queue{Int}()
+    for v in values(automaton.goto[1])
+        enqueue!(q, v)
+    end
+    while !isempty(q)
+        current_state = dequeue!(q)
+        for k in keys(automaton.goto[current_state])
+            new_state = lookup(automaton, current_state, k)
+            enqueue!(q, new_state)
+            state = automaton.fail[current_state]
+            while isnothing(lookup(automaton, state, k))
+                state = automaton.fail[state]
+            end
+            automaton.fail[new_state] = lookup(automaton, state, k)
+            if automaton.output[new_state][1] >
+               automaton.output[automaton.fail[new_state]][1]
+                automaton.output[new_state] = automaton.output[automaton.fail[new_state]] # TODO check if this is the correct way to update output
+            end
+
+        end
+    end
+end
+
+Markdown.@doc doc"""
+    insert_keyword!(aut::AhoCorasickAutomaton, keyword::Word, index::Int)
+
+Insert a new keyword into a given Aho-Corasick automaton to avoid having to rebuild the entire
+automaton.
+""" function insert_keyword!(aut::AhoCorasickAutomaton, keyword::Word, index::Int)
```

Comment:
```suggestion
"""
function insert_keyword!(aut::AhoCorasickAutomaton, keyword::Word, index::Int)
```
The `function` keywords (or `struct` or `mutable struct` etc. etc.) should be on its own line, everywhere